### PR TITLE
foreach => for-each/every, foreach as legacy

### PR DESCRIPTION
**Pull request by hostilefork**
FOREACH was a bit of a black sheep as not using a hyphen
to separate the words, and represented a kind of slippery slope
that had been rejected by Rebol's philosophy (TYPE-OF and not
TYPEOF, for instance).  It also was in a family of other functions
that did have a hyphen, like remove-each and map-each.  It
also draws the eye toward the componet REACH..which once
it is seen sort of can't be unseen.

Debate centered on whether it was common enough to warrant
a non-hyphenated version, and just call it EACH.  That did not
seem to sit well, but the non-hyphenated synonym EVERY came
out as a possibility.  As there are other synonyms, it seems worth
a try.  So speculatively this adds that.

It also keeps FOREACH in the legacy mezzanines.

### CODE REVIEW

**Comment by AntonRolls** [comment]
- I have an objection to hyphenating 'foreach to 'for-each; 
   the legacy is too strong for me. This change would create
   an incompatibility between "old Rebol" and "new Rebol";
   I ask you to imagine all those programmer-hours wasted
   writing compatibility code (which requires someone (ie.
   *you*) to document and maintain it), and of all those
   sorry programmers out there who would have to learn three
   words which all refer to the same function, and also have
   to learn when they can use which words, and make sure
   they have the right version of the compatibility code for
   their version of Rebol...
- I object strongly to the proposed end-goal of there being
   two words (for-each, every) referring to the same
   function; so to read "standard Rebol" would require
   learning that those two words mean the same thing;
   it's a bit of learning that I wouldn't like to do,
   because it's unnecessary.
   It just bloats the documentation a bit, too.
- The objectionableness of the lack of hyphen is not strong
   enough for me. English has plenty of variation with
   hyphenation in composite words, but the sense is that the
   more common the composite word is, the less likely it is
   to have a hyphen (eg. flytrap, mousetrap, mudbrick, ...)
- Rebol draws upon English strongly, and English has plenty
   of irregularities; I think languages have irregularities
   often because words and expressions from obscure/foreign
   backgrounds become common over time, and push their way
   to the front. They are spoken frequently, so tend to
   become a bit shorter and simpler.
   'foreach is very common in Rebol, so I think it deserves
   to be unhyphenated.
- It creates an incompatibility also with Red language.
   Are you going to convince Red language to follow along
   with this change?
   It would require Red to also have documentation and
   compatibility code written and maintained, putting some
   small delays into their development too.
- In summary: I think it is an unnecessary change that
   creates a discontinuity and disturbance, for
   not-strong-enough reasons.
   If it ain't broke, don't fix it!

**Comment by hostilefork** [comment]
@AntonRolls EVERY has been modified as of #83 so it is not a pure synonym.  Instead, it gives back an aggregate result (like a non-short-circuit ALL, e.g. an ALL MAP-EACH).  So you can say, for instance:

    if every x [1 3 5] [odd? x] [
        print "They were all odd"
    ]

This is in contrast to FOR-EACH, which returns the last value the loop gave when it ran (or none if it did not run at all).  Given that the result of "foreach" has likely rarely been used in anyone's code, I imagine pretty much any foreach could still be replaced with EVERY if one wanted to.  But now at least they're different and both have a point.

> This change would create an incompatibility between "old Rebol" and "new Rebol"

The good news I'll give you is that I don't have a particularly strong passion about eliminating the unhyphenated `foreach` from the core distribution.  As long as there's the version that I feel makes sense is there and is metrically better, the less-literate version can stick around.  It costs all of one **foreach: :for-each** so no big rush...and there is no aggressive need to reclaim the unhyphenated foreach for something else.

The bad news is that there are many other things that don't have this character.  They're more foundational, disruptive, or I actively do want to see them killed.  Someday somewhere, we could have a longer discussion about how priorities or decisions are made, how to organize the project, etc.

> so to read "standard Rebol" would require learning that those two words mean the same thing

This specific question aside (with them being different now), I don't know that I would agree on the broader notion that "synonyms are bad"...or that somehow having more than one way to do things is bad because it's more documentation.

You might say that UNLESS is bad because you can make it with IF and NOT, and if we're doing character counts "UNLESS" and "IF NOT" are equal so it's not even any shorter.  But one of Rebol's draws is English-like flow and expression, and synonyms may be part of that.  You sort of have to find out if it makes a difference to users, and how many, etc. 

In any case, your complaints are dismissed.

**Comment by AntonRolls** [comment]
I should apologise because I am a bit behind with the chat, because I started reading older posts a few days ago. I'd only read through to here:
http://chat.stackoverflow.com/transcript/291/2015/8/31
so I've missed later developments with EVERY.
The new behaviour looks like it could be quite useful, so I'm at ease on that front.
You didn't address all my points, but since I've cooled down a little bit, they don't
seem as important, also when taking into consideration all the other stuff that's
changed/changing in Rebol3. At least I've aired them.
So, still not perfectly happy with the slight blurring of the language, but I guess I can live with it.
